### ISSUE TITLE
Concurrency read/write for vault operations

### ISSUE THREAD

**Issue by robert-cronin** [MatrixAI/Polykey#62]
We need to integrate concurrency into our git operations on vaults. Write operations must be FIFO but can be asynchronous to the user. Read operations can be parallel.
With the use of a callback pattern, we can also take advantage of transnational commits. 
Some examples of concurrency libs in node.js are [async-mutex](https://github.com/DirtyHairy/async-mutex) and [live-mutex](https://github.com/ORESoftware/live-mutex). The former is pure typescript. Both can be used with either callbacks or promises. Also, [this](https://medium.com/@the1mills/a-better-mutex-for-node-js-4b4897fd9f11) article provides some background on the latter.

Related to #51

**Comment by robert-cronin** [comment]
I've added the library `async-mutex` for it's pure typescript implementation and it works just find. All I've needed to do is add a `try-catch-finally` block to all git operations that could be parallel write:

```
const release = await this.mutex.acquire()
try {
    ....
} catch (error) {
    release()
    throw error
} finally {
    release()
}
```

I've also added in a test to test concurrency with 50 parallel write operations and they all come out sequential on the git logs.

So this can be closed off with PR #60

**Comment by CMCDragonkai** [comment]
Once this is available, the concept of a "transaction" can be simulated with JS callbacks.

For example the `CounterImmutable` from js-resource-counter does this: https://github.com/MatrixAI/js-resource-counter/blob/master/test/CounterImmutable.js#L129-L136

The callback is supplied with the data structure to do internal mutating ops.

Note that this is a common pattern across all immutable data structures. Sometimes they are called `withMutation ...`. But I like to call them what they are, a "transaction"!.

Note that this could then allow the user to chain operations against a vault, or even between multiple vaults in an atomic operation. But that can be left for later. This is more a internal library pattern for now.
